### ISSUE TITLE
Enhance DIC functionality with new interfaces and traits

### ISSUE THREAD

**Issue by jamisonbryant** [cakephp/cakephp#17558]
### Description

I am a big fan of Dependency Injection and the Service Container, and I've enjoyed taking it for a test run in Cake 4 so far but there are some areas where I've found it a bit awkward to work with.

The biggest example of this is occasionally having to "pass the container around" between objects in order to make those objects able to put things _in_ the container or retrieve them _from_ the container. We solved this in a couple places by adding a constructor arg ([cakephp/authorization#276](https://github.com/cakephp/authorization/pull/276), [cakephp/authorization#278](https://github.com/cakephp/authorization/pull/278), and [cakephp/authentication#662](https://github.com/cakephp/authentication/pull/662)) but this feels a bit bodged-on to me, although I admit it is the best solution we have available at the moment.

What I'd really like to see is an expanded usage of the `ContainerAwareInterface` and the `ContainerAwareTrait` (both present in league/container). Between these two classes developers now have two very powerful tools:

1. They can set up their classes to integrate easily with the application service container without modifying constructors
2. They can provide OR supply a copy of the application service container on-demand to calling classes

This would essentially allow developers to integrate whichever classes they want with the service container, and also allow plugins to offer arbitrary support for fetching objects from the service container without requiring heavy-handed modifications (such as constructor arguments, or unnecessary interfaces) to the underlying classes, making integration (and thus adoption, but that's more opinion than objective fact) easier.

Here's an example of one such workaround I had to create in my code that would have been mitigated by using the `ContainerAware` classes:

1. My app is a Cake 4.x app that uses both official Authz plugins (2.x versions)
2. The Authz plugins were recently updated to be more DI-aware (yay Kevin)
3. My authz service providers are **separate classes**, not methods in the Application class
4. The Authz plugins check for `ContainerApplicationInterface` **but not** `ContainerAwareInterface`
5. Thus, in order to enable DI usage in my service providers I was forced to implement `ContainerApplicationInterface`
6. The end result was that I had one constructor (with one argument, the container) and one noop method (`services()`, required by the interface) that were solely there for the purpose of enabling the Authz plugins to access the container via my service provider classes

Currently, there are PRs open ([cakephp/authorization#278](https://github.com/cakephp/authorization/pull/278) and [cakephp/authentication#662](https://github.com/cakephp/authentication/pull/662)) that add a `$container` argument to the ctors for the Authz middlewares, but honestly I think this would be _way_ better/cleaner:

```
if (
    ($this->subject instanceof ContainerApplicationInterface) ||              // for BC
    ($this->subject instanceof ContainerAwareInterface)                         // the new interface
) {
   // getContainer() available from both interfaces
   // ContainerAwareIface::getContainer() returns a DefinitionContainerIface, which requires add()
   $container = $this->subject->getContainer();
   $container->add(AuthenticationService::class, $service);
}
```

So, now all the plugin knows is that my class provides a container. Then, to use this:

```
// in Application::middleware()
// setContainer()/getContainer() is provided by ContainerAwareTrait,
// which I would just attach to my service providers
$authentication = (new CustomAuthenticationServiceProvider())->setContainer($this->getContainer())
$authorization = (new CustomAuthorizationServiceProvider())->setContainer($this->getContainer());

$middlewareQueue
    ->addMiddleware(new AuthenticationMiddleware($authentication))      // no need to pass container to middleware
    ->addMiddleware(new AuthorizationMiddleware($authorization))
   ...
;
```

To summarize, I think Cake should leverage the built-in league/container classes (interfaces and traits, to be specific) to provide a more seamless, elegant way for userland code and framework code/plugin code to deal with a common service container _without_ doing something crazy like a global container singleton, an idea that has many more drawbacks than positives.

### CakePHP Version

6.0

**Comment by LordSimal** [comment]
I am not that experienced developing such systems but let me just put my thoughts on this one and the rules we have established till now for the DIC (Dependency Injection Container) and more generally the framework:

1. The DIC should only be accessible where its appropriate and necessary. Statically accessing the DIC (like so many other frameworks do) would allow people to shoot themselves in the foot in so many creative and unimaginable ways that its not something we want to think about.
2. We don't have "deep" class structures, meaning there are at most 2 levels of inheritance. Instead we try to use composition over inheritance.

Just those 2 rules heavily restrict our ability to provide the DIC wherever its need but also restrict it to only the things were it makes sense.

I haven't put much thought into this, but a Trait similar to `LocatorAwareTrait` wouldn't restrict accessing the container from anywhere so that won't work. We could of course build a system inside that trait to restrict access to only specific classes but that also seems bad to me.

Having just 1 extra class as a base to hold the container for everything that needs a container also seems pretty stupid and would increase our hierarchy.

So I don't have a real solution to this problem yet as "passing down" is the most straight forward way without compromising the container to the whole framework.

---

As far as I can tell we didn't use the `ContainerAwareTrait` because we just extended the complete implementation of [League\Container\Container](https://github.com/cakephp/cakephp/blob/5.x/src/Core/Container.php#L26)

All that trait provides is functionalitty to get and set the container instance with no fallback to create a new instance (compared to our implementation)

**Comment by rochamarcelo** [comment]
I'm not sure why you need to spam the container and I feel this is not a good way to use the container, based on your code you could simply define the service for `AuthenticationMiddleware`  and use classname when adding the middlewe:

In Application::services

```
$container->add(CustomAuthenticationServiceProvider::class);
$container->add(AuthenticationMiddleware::class)
    ->addArgument(CustomAuthenticationServiceProvider::class);
```

In Application::middleware
``` 
$middlewareQueue->add(AuthenticationMiddleware::class);
```

**Comment by dereuromark** [comment]
Whats the resume here? Should we close this?

**Comment by jamisonbryant** [comment]
There is no direct action that can be taken for the core right now, to my knowledge. I will move some of the above thoughts to a wiki page so that we can preserve them, and we'll continue to debate them until a specific path takes shape.
